app/src/main/java/me/phh/treble/app/HuaweiAudio.kt

```diff
@@ -5,42 +5,51 @@ import android.content.Context
 import android.content.Intent
 import android.content.IntentFilter
 import android.media.AudioManager
+import android.media.AudioSystem
 import android.util.Log
 
 class HuaweiAudio : EntryStartup, BroadcastReceiver() {
     private var audioManager: AudioManager? = null
     override fun onReceive(context: Context, intent: Intent) {
         if (intent.action != Intent.ACTION_HEADSET_PLUG)
             return
-        if (audioManager == null) {
-            audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
-        }
         val state = intent.getIntExtra("state", -1)
         when(state) {
-            0 -> HuaweiHeadsetUtils.speaker(audioManager!!)
-            1 -> HuaweiHeadsetUtils.headset(audioManager!!)
+            AudioSystem.DEVICE_STATE_UNAVAILABLE -> HuaweiHeadsetUtils.speaker(audioManager!!)
+            AudioSystem.DEVICE_STATE_AVAILABLE -> HuaweiHeadsetUtils.headset(audioManager!!)
             else ->
                 Log.e("PlugReceiver", "Unrecognised headset plug state!", Throwable())
         }
         // Apply the changes by setting the volume to the current volume
         // Fails in DND but to exit that you must change the volume later so meh.
-        audioManager!!.setStreamVolume(AudioManager.STREAM_MUSIC, audioManager!!.getStreamVolume(AudioManager.STREAM_MUSIC), 0)
+        val volume = audioManager!!.getStreamVolume(AudioManager.STREAM_MUSIC)
+        audioManager!!.setStreamVolume(AudioManager.STREAM_MUSIC, volume, 0)
     }
 
     override fun startup(ctxt: Context) {
+        audioManager = ctxt.getSystemService(AudioManager::class.java)
         val filter = IntentFilter(Intent.ACTION_HEADSET_PLUG)
         val plugReceiver = this
         ctxt.registerReceiver(plugReceiver, filter)
         Log.d(tag, "Registered for headset plug")
     }
 
+    fun shutdown(ctxt: Context) {
```

Comment:
sounds like you will crash treble_app if you start disabled no?
(by unregistering something that isn't registered)

Reply:
Answered below